pkg/pool-stable-phantom/contracts/StablePhantomPool.sol

```diff
@@ -0,0 +1,466 @@
+    using FixedPoint for uint256;
+    using PriceRateCache for bytes32;
+    using EnumerableMap for EnumerableMap.IERC20ToBytes32Map;
+
+    uint256 private constant _MIN_TOKENS = 2;
+    uint256 private constant _MINIMUM_BPT = 0;
+    uint256 private constant _MAX_TOKEN_BALANCE = 2**(112) - 1;
+
+    uint256 private immutable _bptIndex;
+
+    // Price rate caches are used to avoid querying the price rate for a token every time we need to work with it.
+    // Data is stored with the following structure:
+    //
+    // [   expires   | duration | price rate value ]
+    // [   uint64    |  uint64  |      uint128     ]
+
+    mapping(IERC20 => bytes32) private _priceRateCaches;
+    mapping(IERC20 => IRateProvider) private _rateProviders;
+
+    uint256 private constant _PRICE_RATE_CACHE_VALUE_OFFSET = 0;
+    uint256 private constant _PRICE_RATE_CACHE_DURATION_OFFSET = 128;
+    uint256 private constant _PRICE_RATE_CACHE_EXPIRES_OFFSET = 128 + 64;
+
+    event PriceRateCacheUpdated(IERC20 indexed token, uint256 rate);
+    event PriceRateProviderSet(IERC20 indexed token, IRateProvider indexed provider, uint256 cacheDuration);
+
+    // The constructor arguments are received in a struct to work around stack-too-deep issues
+    struct NewPoolParams {
+        IVault vault;
+        string name;
+        string symbol;
+        IERC20[] tokens;
+        IRateProvider[] rateProviders;
+        uint256[] priceRateCacheDurations;
+        uint256 amplificationParameter;
+        uint256 swapFeePercentage;
+        uint256 pauseWindowDuration;
+        uint256 bufferPeriodDuration;
+        address owner;
+    }
+
+    constructor(NewPoolParams memory params)
+        StablePool(
+            params.vault,
+            params.name,
+            params.symbol,
+            _insertSorted(params.tokens, IERC20(this)),
+            params.amplificationParameter,
+            params.swapFeePercentage,
+            params.pauseWindowDuration,
+            params.bufferPeriodDuration,
+            params.owner
+        )
+    {
+        _require(params.tokens.length >= _MIN_TOKENS, Errors.MIN_TOKENS);
+
+        InputHelpers.ensureInputLengthMatch(
+            params.tokens.length,
+            params.rateProviders.length,
+            params.priceRateCacheDurations.length
+        );
+
+        for (uint256 i = 0; i < params.tokens.length; i++) {
+            _rateProviders[params.tokens[i]] = params.rateProviders[i];
+            _updatePriceRateCache(params.tokens[i], params.rateProviders[i], params.priceRateCacheDurations[i]);
+            emit PriceRateProviderSet(params.tokens[i], params.rateProviders[i], params.priceRateCacheDurations[i]);
+        }
+
+        uint256 bptIndex;
+        for (bptIndex = params.tokens.length; bptIndex > 0 && params.tokens[bptIndex - 1] > IERC20(this); bptIndex--) {
+            // solhint-disable-previous-line no-empty-blocks
+        }
+        _bptIndex = bptIndex;
+    }
+
+    function initialize() external {
```

Comment:
Can you add a note that it is important that this function is called right after creation of the pool and that we recommend to use the factory. In case someone wants to deploy this pool without the factory.